File: 0145-binary-tree-postorder-traversal/0145-binary-tree-postorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        if not root: return []
        res = []
        st = []
        cur = root
        
        while st or cur:
            if cur:
                st.append(cur)
                cur = cur.left
            else:
                tmp = st[-1].right
                if not tmp:
                    tmp = st.pop()
                    res.append(tmp.val)
                    while st and st[-1].right == tmp:
                        tmp = st.pop()
                        res.append(tmp.val)
                else:
                    cur = tmp
                
        return res
        

File: 0145-binary-tree-postorder-traversal/0145-binary-tree-postorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        //postorder(root, result);
        Stack<TreeNode> st = new Stack<>();
        TreeNode cur = root;
        
        while ((!st.isEmpty()) || (cur != null)){
            if (cur != null) {
                st.push(cur);
                cur = cur.left;
            }
            else{
                TreeNode tmp = st.peek().right;
                if (tmp == null){
                    tmp = st.pop();
                    result.add(tmp.val);
                    while ((!st.isEmpty()) && (st.peek().right == tmp)){
                        tmp = st.pop();
                        result.add(tmp.val);
                    }
                }
                else{
                    cur = tmp;
                }
            }
        }
        
        return result;
    }
    
    private void postorder(TreeNode node, List<Integer> result){
        if (node == null) return;
        
        postorder(node.left, result);
        postorder(node.right, result);
        result.add(node.val);
    }
}